### ISSUE TITLE
'exec zsh -l' will open multiple gitstatuses with the same --parent-pid

### ISSUE THREAD

**Issue by docwhat** [romkatv/gitstatus#8]
Running `exec zsh -l` will cause a another `gitstatus` process to appear with the same `--parent-pid`

Running `exec bash -l` will leave any `gitstatus` processes running until the PID stops being used any process.

There is a related bug as well, which is that if `gitstatus` dies (e.g. `kill $(pidof gitstatus)`) then `gitstatus` is never restarted by the shell.

Some off the cuff ideas (some overlap):

- Do locking per parent-pid.
- Have the pipe timeout if not used for a while; have zsh restart gitstatusd if the pipe is missing.
- Use pipe paths not based on (only?) the pid, relying on commands (or lack of) sent via the pipe.
- Only use one gitstatus per user, instead of per-pid?  Spinning up a new one if none is found, using locking to prevent duplicates.

**Comment by romkatv** [comment]
Dear lord! That's a beautiful bug you've found.

I've fixed it in the most straightforward way by ensuring that gitstatusd exits when its parent shell is no more. This was the original intention of the code but I overlooked the fact that a shell can be gone while its process lives on in a different disguise. Now it works. While at it, I also discovered much to my surprise that zsh doesn't close file descriptors >= 10 on exec. Apparently there is no option that changes this. And there I was thinking that zsh has an option for everything. Anyway, I now manually slap `O_CLOEXEC` on all file descriptors.

It's intentional that gitstatusd doesn't get restarted automatically. I want users to come and complain if gitstatusd crashes. Restarts can hide bugs that will eventually catch up with me. It's a bit selfish but I need some of quality/reliability feedback. Disruptive crashes (so far none happened afaik) are still better then telemetry.

So far the approach I took with running an instance of gitstatusd next to each interactive shell has been working well. I'll stick with it until there is a reason to introduce a system-wide daemon.

Once again, thanks for the fantastic bug report. Keep them coming!

P.S.

> Use pipe paths not based on (only?) the pid...

This is already true. All files (pipes, logs, etc.) are generated with mktemp. `$$` is embedded only for debugability.
